Sources/CoreMetrics/Metrics.swift

```diff
@@ -378,6 +382,33 @@ public enum MetricsSystem {
         }
     }
 
+    /// `bootstrapSystemMetrics` is an one-time configuration function which globally enables system level metrics.
+    /// `bootstrapSystemMetrics` can be only called once, unless cancelled usung `cancelSystemMetrics`, calling it more
+    /// than once without cancelling will lead to undefined behaviour, most likely a crash.
+    ///
+    /// - parameters:
+    ///     - pollInterval: The interval at which system metrics should be updated.
+    ///     - systemMetricsType: The type of system metrics to use. If none is provided this defaults to
+    ///                          `LinuxSystemMetrics` on Linux platforms and `NOOPSystemMetrics` on all other platforms.
+    ///     - labels: The labels to use for generated system metrics.
+    public static func bootstrapSystemMetrics(pollInterval interval: DispatchTimeInterval = .seconds(2), systemMetricsType: SystemMetrics.Type? = nil, labels: SystemMetricsLabels) {
```

Comment:
the way I imagined this would work is a separate module that extends `MetricsSystem` and adds a new variant of `bootstrap`, something like: 

```
public static func bootstrap(_ factory: MetricsFactory, systemMetrics: SystemMetricsConfiguration) {
```
having  SystemMetricsConfiguration a container for the `pollInternval` and other configuration options

Reply:
I'd like to do that, but because of the access modifiers in `MetricsSystem` this is not possible. I have managed to lose the `Foundation` dependency, so there might not be a reason anymore to split out?

Reply:
@MrLotU what about something like: 

```
import CoreMetrics
import Dispatch

extension MetricsSystem {
    private static var systemMetricsFactory: SystemMetricsFactory?
    
    public static func bootstrapWithSystemMetrics(_ factory: MetricsFactory, configuration: SystemMetrics.Configuration) {
        let factory = SystemMetricsFactory(factory: factory, configuration: configuration)
        self.bootstrap(factory)
        systemMetricsFactory = factory
        systemMetricsFactory!.poll()
    }
    
    public static func cancelSystemMetrics() {
        systemMetricsFactory?.cancelPolling()
    }
    
    internal struct SystemMetricsFactory: MetricsFactory {
        let underlying: MetricsFactory
        let configuration: SystemMetrics.Configuration

        let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsHandler", qos: .background)
        let task: DispatchWorkItem
        
        init (factory: MetricsFactory, configuration: SystemMetrics.Configuration) {
            self.underlying = factory
            self.configuration = configuration
            self.task = DispatchWorkItem(qos: .background, block: {
                let metrics = SystemMetricsFactory.readSystemMetrics()
                Gauge(label: configuration.labels.label(for: \.virtualMemoryBytes)).record(metrics.virtualMemoryBytes)
                Gauge(label: configuration.labels.label(for: \.residentMemoryBytes)).record(metrics.residentMemoryBytes)
                Gauge(label: configuration.labels.label(for: \.startTimeSeconds)).record(metrics.startTimeSeconds)
                Gauge(label: configuration.labels.label(for: \.cpuSecondsTotal)).record(metrics.cpuSeconds)
                Gauge(label: configuration.labels.label(for: \.maxFds)).record(metrics.maxFds)
                Gauge(label: configuration.labels.label(for: \.openFds)).record(metrics.openFds)
            })
        }
        
        func poll() {
            self.queue.asyncAfter(deadline: .now() + self.configuration.interval) {
                guard !self.task.isCancelled else {
                    return
                }
                self.task.perform()
                self.poll()
            }
        }
        
        func cancelPolling() {
            self.task.cancel()
        }
        
        // MARK: - proxy
        
        func makeCounter(label: String, dimensions: [(String, String)]) -> CounterHandler {
            self.underlying.makeCounter(label: label, dimensions: dimensions)
        }
        
        func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> RecorderHandler {
            self.underlying.makeRecorder(label: label, dimensions: dimensions, aggregate: aggregate)
        }
        
        func makeTimer(label: String, dimensions: [(String, String)]) -> TimerHandler {
            self.underlying.makeTimer(label: label, dimensions: dimensions)
        }
        
        func destroyCounter(_ handler: CounterHandler) {
            self.underlying.destroyCounter(handler)
        }
        
        func destroyRecorder(_ handler: RecorderHandler) {
            self.underlying.destroyRecorder(handler)
        }
        
        func destroyTimer(_ handler: TimerHandler) {
            self.underlying.destroyTimer(handler)
        }
        
        // MARK: - system metrics
        
        private static func readSystemMetrics() -> SystemMetrics.Data {
            fatalError("FIXME")
        }
    }
    
}

public enum SystemMetrics {
        
    public struct Configuration {
        let interval: DispatchTimeInterval
        let labels: Labels
        let metricsType: SystemMetrics.Type?
        
        public init(pollInterval interval: DispatchTimeInterval = .seconds(2),
                    labels: Labels,
                    metricsType: SystemMetrics.Type? = nil) {
            self.interval = interval
            self.labels = labels
            self.metricsType = metricsType
            
        }
    }
    
    public struct Labels {
        let prefix: String
        let virtualMemoryBytes: String
        let residentMemoryBytes: String
        let startTimeSeconds: String
        let cpuSecondsTotal: String
        let maxFds: String
        let openFds: String

        public init(prefix: String, virtualMemoryBytes: String, residentMemoryBytes: String, startTimeSeconds: String, cpuSecondsTotal: String, maxFds: String, openFds: String) {
            self.prefix = prefix
            self.virtualMemoryBytes = virtualMemoryBytes
            self.residentMemoryBytes = residentMemoryBytes
            self.startTimeSeconds = startTimeSeconds
            self.cpuSecondsTotal = cpuSecondsTotal
            self.maxFds = maxFds
            self.openFds = openFds
        }

        func label(for keyPath: KeyPath<SystemMetrics.Labels, String>) -> String {
            return self.prefix + self[keyPath: keyPath]
        }
    }
    
    public struct Data {
        let pid: Int
        let virtualMemoryBytes: Int
        let residentMemoryBytes: Int
        let startTimeSeconds: Int
        let cpuSeconds: Int
        let maxFds: Int
        let openFds: Int
    }
}
```

have not tested it but I think can be made to work